[TensorExpr] Add IRVerifier.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52901 [TensorExpr] Add IRVerifier.**

This PR implements IR Verifier and adds a call to it in `LoopNest`
constructors. Checks that were in expr/stmt constructors before are now
moved to the corresponding `::make` functions or to the verifier. They
didn't really help from the constructors anyway since an exception
thrown from there led to a segfault due to the fact our memory
management works (object was not fully created but was registered in the
kernel arena for destruction anyway).

Fixes #52778.

Differential Revision: [D26682928](https://our.internmc.facebook.com/intern/diff/D26682928)

Commit:
[TensorExpr] Add IRVerifier.

This PR implements IR Verifier and adds a call to it in `LoopNest`
constructors. Checks that were in expr/stmt constructors before are now
moved to the corresponding `::make` functions or to the verifier. They
didn't really help from the constructors anyway since an exception
thrown from there led to a segfault due to the fact our memory
management works (object was not fully created but was registered in the
kernel arena for destruction anyway).

Fixes #52778.

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/tensorexpr/ir.h:
```diff
@@ -485,22 +482,21 @@ class IfThenElse : public ExprNode<IfThenElse> {
-    if (t->dtype() != f->dtype()) {
+    if (t.dtype() != f.dtype()) {
       throw malformed_input("Bad dtype in IfThenElse");
     }
```
These have already been moved to the verifier, right? Why do we need them here as well?

ZolotukhinM replied:
I kept them in `::make` methods whenever it was possible, since there they don't cause segfaults and will allow to catch the issue earlier.

navahgar reviewed (approved):
Looks much better now. Thanks for moving them all together. It will be very useful to run this verifier pass wherever we need it.

How about explicit unit tests for the verifier? If that is going to be a lot of work, we can do that later.

ZolotukhinM commented:
> It will be very useful to run this verifier pass wherever we need it.

Yeah, absolutely agree. In future we might have different verifiers as well (e.g. one that would check that all buffers has `Allocate` for them - but that could only be run after `Allocate`s had been inserted). I listed some other ideas in #52583.

> How about explicit unit tests for the verifier?

They might be useful, let me add some.

Commit:
Update on "[TensorExpr] Add IRVerifier."


This PR implements IR Verifier and adds a call to it in `LoopNest`
constructors. Checks that were in expr/stmt constructors before are now
moved to the corresponding `::make` functions or to the verifier. They
didn't really help from the constructors anyway since an exception
thrown from there led to a segfault due to the fact our memory
management works (object was not fully created but was registered in the
kernel arena for destruction anyway).

Fixes #52778.

Differential Revision: [D26682928](https://our.internmc.facebook.com/intern/diff/D26682928)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IRVerifier."


This PR implements IR Verifier and adds a call to it in `LoopNest`
constructors. Checks that were in expr/stmt constructors before are now
moved to the corresponding `::make` functions or to the verifier. They
didn't really help from the constructors anyway since an exception
thrown from there led to a segfault due to the fact our memory
management works (object was not fully created but was registered in the
kernel arena for destruction anyway).

Fixes #52778.

Differential Revision: [D26682928](https://our.internmc.facebook.com/intern/diff/D26682928)

[ghstack-poisoned]
